### PR TITLE
Image

### PR DESCRIPTION
**Pull request by graphemecluster**
Thank you for letting me open this pull request. I don't mind if you commit again yourself in case you don't want my name to appear in your repository.

### CODE REVIEW

**Comment by graphemecluster** [comment]
I think you should look for another free (better open-source) iOS application doing the same job, since there are only ARKit parameters. Some users are not willing to pay money, you know.

File: tha2/app/ifacialmocap_puppeteer.py
Comment (dragonmeteor) [inline tha2/app/ifacialmocap_puppeteer.py:395]: It seemed that you have removed the functionalities here. Please preserve them.

File: tha2/app/ifacialmocap_puppeteer.py
Comment (dragonmeteor) [inline tha2/app/ifacialmocap_puppeteer.py:406]: To be more specific, [lease reinstate the background insertion.

File: tha2/util.py
Comment (dragonmeteor) [inline tha2/util.py:142]: Have you thought about the case where the user opens a file without an alpha channel? 

I believe the alpha channel would always be defaulted to 1, and the output would be ugly even with your zeroing preprocessing. Here's an example: https://imgur.com/6eifxWR.

I don't think it's a good idea to prematurely convert the input image to RGBA. Please reinstate the channel checking code.

File: tha2/util.py
Comment (dragonmeteor) [inline tha2/util.py:151]: Do not remove this line.

File: tha2/app/manual_poser.py
Comment (dragonmeteor) [inline tha2/app/manual_poser.py:277]: Please preserve the channel count checking. (See the comment on converting PIL image to RGBA below).

**Comment by graphemecluster** [comment]
I did think about all cases. I admit that I did this intentionally.
How about if the user opens a file with an alpha channel but with no alpha at all (all values on alpha channel are 1)?

A better method would be to implement chroma key.

Anyway, I would add back the alpha channel checking. Sorry for any inconvenience caused.

**Comment by graphemecluster** [comment]
Sorry, I accidentally remove the background blending part when I change the method to `convert_output_image_from_torch_to_numpy`.
Originally I am going to remove the `blend_with_background` method because Windows does support virtual camera streaming with alpha channel.

File: tha2/app/ifacialmocap_puppeteer.py
Comment (dragonmeteor) [inline tha2/app/ifacialmocap_puppeteer.py:438]: Change to w,h = pil_image.size. For a PIL image, size is a 2-tuple, not a 3-tuple. (See https://pillow.readthedocs.io/en/4.1.x/reference/Image.html#PIL.Image.size)

File: tha2/app/ifacialmocap_puppeteer.py
Comment (dragonmeteor) [inline tha2/app/ifacialmocap_puppeteer.py:439]: Change to:

pil_image.mode != 'RGBA':

File: tha2/app/manual_poser.py
Comment (dragonmeteor) [inline tha2/app/manual_poser.py:267]: Ditto.

File: tha2/app/manual_poser.py
Comment (dragonmeteor) [inline tha2/app/manual_poser.py:268]: Ditto.

**Review by dragonmeteor — CHANGES_REQUESTED** [review]
The reinstated channel check did not conform to PIL Image's interface, so image loading did not work at all. Please fix as I commented.

File: tha2/app/manual_poser.py
Comment (dragonmeteor) [inline tha2/app/manual_poser.py:329]: Please also import numpy at the begining of the filed.

**Review by dragonmeteor — CHANGES_REQUESTED** [review]

**Comment by graphemecluster** [comment]
Oh sorry, yesterday I altered it in a hurry and I didn't test it at all.
But actually I assumed you to edit them and commit to my branch before merging.

File: tha2/util.py
Comment (dragonmeteor) [inline tha2/util.py:208]: I've discovered a new problem: the Jupyter notebooks (tha2.ipynb and the other one) stopped working after this line is removed. Here's what it looks like: https://imgur.com/JPQwk2g

The code in the Jupyter notebook expects the output of this function to type dtype=uint8. Without this line, it becomes dtype=float32, which results in the image not being converted in the right way.

Please revert the removal of this line.

File: tha2/app/ifacialmocap_puppeteer.py
Comment (dragonmeteor) [inline tha2/app/ifacialmocap_puppeteer.py:394]: (See the comment on convert_output_image_from_torch_to_numpy  first.) 

Now that convert_output_image_from_torch_to_numpy will output an uint8 tensor, it will not work with the statements below. Please remove this line and replace it with the if statement that you removed earlier:

        if output_image.shape[0] == 4:
            numpy_image = rgba_to_numpy_image(output_image)
        elif output_image.shape[0] == 1:
            c, h, w = output_image.shape
            alpha_image = torch.cat([output_image.repeat(3, 1, 1) * 2.0 - 1.0, torch.ones(1, h, w)], dim=0)
            numpy_image = rgba_to_numpy_image(alpha_image)
        elif output_image.shape[0] == 2:
            numpy_image = grid_change_to_numpy_image(output_image, num_channels=4)
        else:
            raise RuntimeError("Unsupported # image channels: " + output_image.shape[0])

File: tha2/app/manual_poser.py
Comment (dragonmeteor) [inline tha2/app/manual_poser.py:330]: Now that convert_output_image_from_torch_to_numpy will output a uint8 tensor, this conversion is not needed. Please revert.

**Review by dragonmeteor — CHANGES_REQUESTED** [review]

**Comment by dragonmeteor** [comment]
Let me take this opportunity to answer some questions and comments:

> I think you should look for another free (better open-source) iOS application doing the same job, since they are only pure ARKit parameters. Some users are not willing to pay money, you know.
Thank you for the suggestion, but I simply have no motivation to do this right now.

> How about if the user opens a file with an alpha channel but with no alpha at all (all values on alpha channel are 1)?
There's already a specification on how to prepare input images written in the README file. In this case, the image does not meet the specification, so the output is undefined. It is hard to check whether all the pixels have "correct" alpha values. However, it is much easier to check whether the image has an alpha channel at all. So, I think it is a good idea to tell the user of the mistake when it is easy to do so.

> A better method would be to implement chroma key.
To keep the features of this demo minimal, I currently do not plan to implement chroma key or include it as a feature.

> But actually I assumed you to edit them and commit to my branch before merging.
I believe this is your pull request. It would not be respectful to you if I were to interfere with it.

I believe the pull request is close to being done. Just only a few more bugs to fix. Please bear with me for a little bit more.

**Comment by graphemecluster** [comment]
Actually that's because I totally forgot about the notebooks – It works great after alternations.
Some lines changed while exporting the notebooks (because the `.ipynb` format was updated?), but I changed back a little bit to conform with yours.

Other minor changes include:
- The accelerator is set to GPU now.
- `torch` is imported at last because I got this error: `ValueError: module functions cannot set METH_CLASS or METH_STATIC`.

**Review by dragonmeteor — APPROVED** [review]
LGTM

**Comment by dragonmeteor** [comment]
Please commit. Thank you for your hard work.

**Comment by graphemecluster** [comment]
No, I should be thanking you.
… Um, I don't think I have write access to this repository.

**Comment by pkhungurn** [comment]
Ah, sorry. I've merged it myself.

**Comment by graphemecluster** [comment]
I forgot to also remove the FAQ part from README.md.

**Pull request merged**
